src/vm/database/structures.rs

```diff
@@ -141,44 +152,124 @@ type Result<T> = std::result::Result<T, Error>;
+                amount_locked,
+                unlock_height,
+            } => {
+                buffer
+                    .write_all(&[STXBalance::pox_2_version])
+                    .expect("STXBalance serialization: failed to write PoX version byte");
+                buffer
+                    .write_all(&amount_unlocked.to_be_bytes())
+                    .expect("STXBalance serialization: failed writing amount_unlocked.");
+                buffer
+                    .write_all(&amount_locked.to_be_bytes())
+                    .expect("STXBalance serialization: failed writing amount_locked.");
+                buffer
+                    .write_all(&unlock_height.to_be_bytes())
+                    .expect("STXBalance serialization: failed writing unlock_height.");
+            }
+        }
         to_hex(buffer.as_slice())
     }
 }
 
 impl ClarityDeserializable<STXBalance> for STXBalance {
     fn deserialize(input: &str) -> Self {
         let bytes = hex_bytes(&input).expect("STXBalance deserialization: failed decoding bytes.");
-        assert_eq!(bytes.len(), STXBalance::size_of);
-
-        let amount_unlocked = u128::from_be_bytes(
-            bytes[0..16]
-                .try_into()
-                .expect("STXBalance deserialization: failed reading amount_unlocked."),
-        );
-        let amount_locked = u128::from_be_bytes(
-            bytes[16..32]
-                .try_into()
-                .expect("STXBalance deserialization: failed reading amount_locked."),
-        );
-        let unlock_height = u64::from_be_bytes(
-            bytes[32..40]
-                .try_into()
-                .expect("STXBalance deserialization: failed reading unlock_height."),
-        );
-
-        STXBalance {
-            amount_unlocked,
-            amount_locked,
-            unlock_height,
+        if bytes.len() == STXBalance::unlocked_and_v1_size {
+            let amount_unlocked = u128::from_be_bytes(
+                bytes[0..16]
+                    .try_into()
+                    .expect("STXBalance deserialization: failed reading amount_unlocked."),
+            );
+            let amount_locked = u128::from_be_bytes(
+                bytes[16..32]
+                    .try_into()
+                    .expect("STXBalance deserialization: failed reading amount_locked."),
+            );
+            let unlock_height = u64::from_be_bytes(
+                bytes[32..40]
+                    .try_into()
+                    .expect("STXBalance deserialization: failed reading unlock_height."),
+            );
+
+            if amount_locked == 0 {
+                STXBalance::Unlocked {
+                    amount: amount_unlocked,
+                }
+            } else {
+                STXBalance::LockedPoxOne {
+                    amount_unlocked,
+                    amount_locked,
+                    unlock_height,
+                }
+            }
+        } else if bytes.len() == STXBalance::v2_size {
```

Comment:
Since we're probably going to be versioning this going forward, want to just make this into something like this?

```
} else {
   assert_eq!(bytes.len(), STXBalance::v2_size, "Unrecognized serialized STXBalance");
   let version = bytes[0];
   match version {
      STXBalance::pox_2_version => {
          // ...
      }
      _ => {
          // add more versions as supported
          panic!("Bad version byte in STXBalance serialization = {}", version);
       }
    }
 }
 ```

Reply:
But that code structure assumes that the size would remain the same in version 3, etc., which isn't necessarily case, so I'd prefer a code structure that fits more closely with the current check (i.e., without a single case match).

Reply:
Sounds good